src/hal/stm32f4/pin.rs

```diff
@@ -151,21 +124,103 @@ impl PinConf {
   }
 }
 
+impl Gpio for Pin {
+  /// Sets output GPIO value to high.
+  fn set_high(&self) {
+    let offset = self.pin as usize;
+    self.get_reg().bsrr.set_bs(offset, true);
+  }
+
+  /// Sets output GPIO value to low.
+  fn set_low(&self) {
+    let offset = self.pin as usize;
+    self.get_reg().bsrr.set_br(offset, true);
+  }
+
+  /// Returns input GPIO level.
+  fn level(&self) -> GpioLevel {
+    let offset = self.pin as usize;
+    let reg = self.get_reg();
+
+    match reg.idr.id(offset) {
+      false => GpioLevel::Low,
+      _ => GpioLevel::High,
+    }
+  }
+
+  /// Sets output GPIO direction.
+  fn set_direction(&self, new_mode: GpioDirection) {
+    // TODO(darayus): Verify that this works
+    // TODO(darayus): Change the Pin.function field to the new mode
+    use self::reg::GPIO_moder_mode as RegMode;
+    let offset = self.pin as usize;
+    let reg = self.get_reg();
+
+    let val = match new_mode {
+      GpioDirection::Out => RegMode::Output,
+      GpioDirection::In  => RegMode::Input,
+    };
+
+    reg.moder.set_mode(offset, val);
+  }
+}
+
 #[allow(dead_code)]
 mod reg {
+  use core::ops::Drop;
   use volatile_cell::VolatileCell;
 
-  ioreg_old!(GPIO: u32, MODER, OTYPER, OSPEEDER, PUPDR, IDR, ODR, BSRR, LCKR, AFRL, AFRH);
-  reg_rw!(GPIO, u32, MODER,    set_MODER,    MODER);
-  reg_rw!(GPIO, u32, OTYPER,   set_OTYPER,   OTYPER);
-  reg_rw!(GPIO, u32, OSPEEDER, set_OSPEEDER, OSPEEDER);
-  reg_rw!(GPIO, u32, PUPDR,    set_PUPDR,    PUPDR);
-  reg_rw!(GPIO, u32, IDR,      set_IDR,      IDR);
-  reg_rw!(GPIO, u32, ODR,      set_ODR,      ODR);
-  reg_rw!(GPIO, u32, BSRR,     set_BSRR,     BSRR);
-  reg_rw!(GPIO, u32, LCKR,     set_LCKR,     LCKR);
-  reg_rw!(GPIO, u32, AFRL,     set_AFRL,     AFRL);
-  reg_rw!(GPIO, u32, AFRH,     set_AFRH,     AFRH);
+  ioregs!(GPIO = {
```

Comment:
Awesome work @posborne! I wonder if we should keep SVDs in zinc tree, there are pros and cons to that. We need to modify some of them to make them at least working (LPC files have some constant names trimmed), but them we can clean them up for the sake of readability too.